src/core/common/heap_allocatable.hpp

```diff
@@ -0,0 +1,116 @@
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for `Heap::Allocatable`.
+ */
+
+#ifndef HEAP_ALLOCATABLE_HPP_
+#define HEAP_ALLOCATABLE_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/code_utils.hpp"
+#include "common/error.hpp"
+#include "common/heap.hpp"
+#include "common/new.hpp"
+
+namespace ot {
+namespace Heap {
+
+/**
+ * This template class defines a `Heap::Allocatable` object.
+ *
+ * `Heap::Allocatable` provides `New()` and `Free()` methods to allocate and free instances of `Type` on heap.
+ *
+ * Users of this class should follow CRTP-style inheritance, i.e., the `Type` class itself should inherit from
+ * `Allocatable<Type>`.
+ *
+ * The `Type` class destructor is used when `Free()` is called. The destructor frees any heap allocated data members
+ * that are stored in a `Type` instance (e.g., `Heap::String`, `Heap::Data`, etc).
+ *
+ */
+template <class Type> class Allocatable
+{
+public:
+    /**
+     * This static method allocates a new instance of `Type` on heap and initializes it.
+     *
+     * The `Type class MUST have a default constructor (with no arguments) which is invoked upon allocation of new
+     * `Type` instance. It MUST also provide an `Error Init(Args...)` method to initialize the instance. If any `Error`
+     * other than `kErrorNone` is returned the initialization is considered failed.
+     *
+     * @param[in] aArgs   A set of arguments to initialize the allocated `Type` instance.
+     *
+     * @returns A pointer to the newly allocated instance or `nullptr` if it fails to allocate or initialize.
+     *
+     */
+    template <typename... Args> static Type *New(Args &&... aArgs)
+    {
+        void *buf    = Heap::CAlloc(1, sizeof(Type));
+        Type *object = nullptr;
+
+        VerifyOrExit(buf != nullptr);
+
+        object = new (buf) Type();
+
+        if (object->Init(static_cast<Args &&>(aArgs)...) != kErrorNone)
```

Comment:
I think it's better log the error, or we can just return a boolean for the `Init` method. Thoughts?

Reply:
The `Init()` on `Type` class may be used on its own (i.e., on a not heap allocated instance). Returning `Error` gives us more flexibility in such a case.